mujoco_py/tests/test_pid.py

```diff
@@ -44,51 +48,98 @@
+    refer to the setting in the PID_ACTUATOR.
 
-	# pertubation of pole to be unbalanced
-	init_pos = 0.1 * (random.random() - 0.5)
-	print('init pos', init_pos)
-	sim.data.qpos[0] = init_pos
+    Here we set Kp = 200, Ti = 10, Td = 0.1 (also iClamp = 10.0, dSmooth be 0.1)
+    """
+    random.seed(30)
+    xml = MODEL_XML.format(actuator=PID_ACTUATOR, nuser_actuator=1)
+    model = load_model_from_xml(xml)
+    sim = MjSim(model)
+    cymj.set_pid_control(sim.model, sim.data)
 
-	pos = 0.0
-	sim.data.ctrl[0] = pos
-	print('desire position:', pos)
+    # pertubation of pole to be unbalanced
+    init_pos = 0.1 * (random.random() - 0.5)
+    print('init pos', init_pos)
+    sim.data.qpos[0] = init_pos
 
-	for _ in range(100):
-		sim.step()
+    pos = 0.0
+    sim.data.ctrl[0] = pos
+    print('desired position:', pos)
 
-	print('final pos', sim.data.qpos[0])
-	assert abs(sim.data.qpos[0] - pos) < 0.01
+    for _ in range(1000):
+        sim.step()
 
-"""
+    print('final pos', sim.data.qpos[0])
+    assert abs(sim.data.qpos[0] - pos) < 1e-4
+
+
+def test_mj_proportional_only():
+    """
     check new PID control is backward compatible with  position control
-	when only has Kp term.
-"""
-def test_mj_proptional_only():
-	model = load_model_from_xml(MODEL_XML.format(actuator=P_ONLY_ACTUATOR))
-	sim = MjSim(model)
-	cymj.set_pid_control(sim.model, sim.data)
-
-	model2 = load_model_from_xml(MODEL_XML.format(actuator=POSITION_ACTUATOR))
-	sim2 = MjSim(model2)
-
-	init_pos = 0.1 * (random.random() - 0.5)
-	sim.data.qpos[0] = sim2.data.qpos[0] = init_pos
-	sim.data.ctrl[0] = sim2.data.ctrl[0] = 0
-
-	for i in range(2000):
-		print(i, sim.data.qpos[0], sim2.data.qpos[0])
-		sim.step()
-		sim2.step()
-		assert abs(sim.data.qpos[0] - sim2.data.qpos[0]) <= 1e-7, "%d step violates" % i
+    when only has Kp term.
+    """
+    model = load_model_from_xml(MODEL_XML.format(actuator=P_ONLY_ACTUATOR, nuser_actuator=1))
+    sim = MjSim(model)
+    cymj.set_pid_control(sim.model, sim.data)
+
+    model2 = load_model_from_xml(MODEL_XML.format(actuator=POSITION_ACTUATOR, nuser_actuator=1))
+    sim2 = MjSim(model2)
+
+    init_pos = 0.1 * (random.random() - 0.5)
+    sim.data.qpos[0] = sim2.data.qpos[0] = init_pos
+    sim.data.ctrl[0] = sim2.data.ctrl[0] = 0
+
+    for i in range(2000):
+        print(i, sim.data.qpos[0], sim2.data.qpos[0])
+        sim.step()
+        sim2.step()
+        assert abs(sim.data.qpos[0] - sim2.data.qpos[0]) <= 1e-7, "%d step violates" % i
+
+
+def test_cascaded_pid():
+    """
+    To enable Cascaded PID control in the mujoco, please
+    refer to the setting in the CASCADED_PID_ACTUATOR. user param should be set to 1
+
+    Here we set Kp = 5 for the position control loop and Kp =  10 for the velocity control
+    Ti = 0.1 and integral_max_clamp=1.5.
+    EMA smoothing constant is set to 0.97, and velocity limit is 3 rad/s
+    """
+    random.seed(30)
+    xml = MODEL_XML.format(actuator=CASCADED_PID_ACTUATOR, nuser_actuator=1)
```

Comment:
shall we add some tests to compare cascaded pid with regular pid to check the expected benefits?? e.g. the max velocity should be smaller or something.

Reply:
added one for the cascaded PID. I can also extend the existing PID test (now that I set seed, we could ideally test for trajectories, but didn't want to do that here since it's quite a lot of data)

Reply:
what is the max_torque for pid only controller here?

Reply:
adding @rubendsa: can we set a PID with only PI terms set, and compared against cascaded with extra velocity constraints, is that a unit test we could add?

Reply:
@ilge Yeah, we can run that test. In this context with both loops running at the same frequency, the biggest gain we have with this cascaded approach is having the ability to apply velocity limits independent of torque constraints.

Reply:
I couldn't find a good way to test PIDs that have similar Kp/Ti terms for the position control as per for the cascade case, the position controller part is acting to trace a velocity setpoint ant it'll inherently have a very different transfer function compared to the single PID. let me know if the individual tests don't look sufficient, though.